src/rpc/client/frame.ts

```diff
@@ -75,14 +75,61 @@ export class Frame extends ChannelOwner<FrameChannel, FrameInitializer> {
       if (event.remove)
         this._loadStates.delete(event.remove);
     });
+    this._channel.on('navigated', event => {
+      this._url = event.url;
+      this._name = event.name;
+      this._eventEmitter.emit('navigated', event);
+      if (!event.error && this._page)
+        this._page.emit(Events.Page.FrameNavigated, this);
+    });
   }
 
   async goto(url: string, options: GotoOptions = {}): Promise<network.Response | null> {
     return network.Response.fromNullable((await this._channel.goto({ url, ...options, isPage: this._page!._isPageCall })).response);
   }
 
+  private _setupNavigationWaiter(): Waiter {
+    const waiter = new Waiter();
+    waiter.rejectOnEvent(this._page!, Events.Page.Close, new Error('Navigation failed because page was closed!'));
+    waiter.rejectOnEvent(this._page!, Events.Page.Crash, new Error('Navigation failed because page crashed!'));
+    waiter.rejectOnEvent<Frame>(this._page!, Events.Page.FrameDetached, new Error('Navigating frame was detached!'), frame => frame === this);
+    return waiter;
+  }
+
   async waitForNavigation(options: types.WaitForNavigationOptions = {}): Promise<network.Response | null> {
-    return network.Response.fromNullable((await this._channel.waitForNavigation({ ...options, isPage: this._page!._isPageCall })).response);
+    const waitUntil = verifyLoadState(options.waitUntil === undefined ? 'load' : options.waitUntil);
+    const timeout = this._page!._timeoutSettings.navigationTimeout(options);
+    const apiName = this._page!._isPageCall ? 'page.waitForNavigation' : 'frame.waitForNavigation';
+    const waiter = this._setupNavigationWaiter();
+    waiter.rejectOnTimeout(timeout, new TimeoutError(`Timeout ${timeout}ms exceeded during ${apiName}.`));
+
+    const toUrl = typeof options.url === 'string' ? ` to "${options.url}"` : '';
+    waiter.log(`waiting for navigation${toUrl} until "${waitUntil}"`);
+
+    const navigatedEvent = await waiter.waitForEvent<FrameNavigatedEvent>(this._eventEmitter, 'navigated', event => {
+      // Any failed navigation results in a rejection.
+      if (event.error)
+        return true;
+      waiter.log(`  navigated to "${event.url}"`);
+      return helper.urlMatches(event.url, options.url);
```

Comment:
What about the predicate version?

Reply:
`options.url` is a `types.URLMatch` object that contains a predicate, so we are good :)